src/pybind/mgr/dashboard/controllers/smb.py

```diff
@@ -0,0 +1,94 @@
+
+# -*- coding: utf-8 -*-
+
+import logging
+import json
+
+from typing import Dict, List, Union
+
+
+from dashboard.controllers._permissions import CreatePermission, DeletePermission
+from smb import results
+from smb.enums import AuthMode, SMBClustering, Intent
+
+from .. import mgr
+from ..security import Scope
+from . import APIDoc, APIRouter, ReadPermission, RESTController
+
+logger = logging.getLogger('controllers.smb')
+
+
+@APIRouter('/smb/cluster', Scope.SMB)
+@APIDoc("SMB Cluster Management API", "SMB")
+class SMBCluster(RESTController):
+    _resource: str = 'ceph.smb.cluster'
+
+    @ReadPermission
+    def list(self) -> List[Dict]:
+        return mgr.remote('smb', 'show', [self._resource])
+
+    @CreatePermission
+    def create(
+        self,
+        cluster_id: str,
+        auth_mode: AuthMode,
+        intent: Intent = Intent.PRESENT,
+        domain_settings: Dict = {},
+        user_group_settings: List[str] = [],
+        custom_dns: List[str] = [],
+        placement: str = '',
+        clustering: Union[SMBClustering, None] = None,
+        public_addrs: List[str] = [],
+        custom_smb_global_options: Dict = {}
+    ) -> results.Simplified:
+        # pylint: disable=unused-argument
+        cluster_data = locals().items()
+
+        resource = {}
+        resource['resource_type'] = self._resource
+
+        for key, value in cluster_data:
+            if key != 'self' and value:
+                if key == 'domain_settings' and domain_settings:
+                    if 'domain_settings' not in resource:
+                        resource[key] = {}
+                        resource[key]['join_sources'] = []
+
+                    resource[key]['realm'] = domain_settings['realm'] if 'realm' in domain_settings else None
+
+                    for auth_id in domain_settings['join_sources'] if 'join_sources' in domain_settings else []:
+                        value = {'source_type': 'resource', 'ref': auth_id}
+                        resource[key]['join_sources'].append(value)
+
+                elif key == 'user_group_settings' and user_group_settings:
+                    if 'user_group_settings' not in resource:
+                            resource[key] = []
+
+                    for user_id in user_group_settings:
+                        value = {'source_type': 'resource', 'ref': user_id}
+                        resource[key].append(value)
+                else:
+                    resource[key] = value
+        res =  mgr.remote('smb', 'apply_resources', json.dumps(resource))
+        return res.to_simplified() if type(res) is results.Result else res
```

Comment:
You are right @nizamial09, maybe I went over the top with this... But let me explain why:
I though it would be easier for the user to work with the fields or query params than an opaque json object where you would need to make us of "resources" structures, so the current implementation tries to abstract the "resource" entity that uses the smb mgr module.

So, for the current implementation you can use a json like:
```
{
  "cluster_id": "rhumba",
  "auth_mode": "user",
  "user_group_settings": [ "ug1" ]
}
```
We could do as you suggested of course, and simply input a json obj:
```
    def create(
        self,
        cluster_data: Dict
    ) -> results.Simplified:
        return mgr.remote('smb', 'apply_resources', json.dumps(cluster_data)).to_simplified()
```

But for it to work we would need something like:
```
{
  "cluster_data": {
    "resource_type": "ceph.smb.cluster",
    "cluster_id": "rhumba",
    "auth_mode": "user",
    "user_group_settings": [
      {
        "source_type": "resource",
        "ref": "ug1"
      }
    ]
}
```
I though there was no need to request the `resource_type` , `source_type` fields, etc... So this implementation tries to abstract that.

So if we don't want to abstract anything I suppose that's okay, but I personally don't like the idea of having actions as "cluster create" in this case have to deal with resources.

What do you think @nizamial09?

Reply:
I'd prefer simplicity here and probably for a request that is aligned with the CLI. So I think having the same json won't be too bad. It'd be future proof as well since we don't need to adapt a lot of whatever change CLI is making to it. Also, this is what we already do on the osd and service creation. UI has many fields which later generates a json and send it to backend. So I think we can start there, and also in the `EndpointDoc` we can mention the format of the req. If needed, later on we can see whatever customization we need. WDYT?

Reply:
> for a request that is aligned with the CLI
I get that, since that's what we usually do in the Dashboard but, this is not aligned with the CLI either. The CLI cmd does not deal with resources either, CLI asks for `domain_join_user_pass`/`define_user_pass` which is a string in the form <username>%<password>  instead of `user_group_settings`/`domain_settings` which are resource objects. 
Source: https://docs.ceph.com/en/latest/mgr/smb/#create-cluster

> So I think having the same json won't be too bad.
So yes, we will have the same available json which I agree it's okay, but this is not the behaviour of the `smb cluster` cmds as said before.

>  and also in the EndpointDoc we can mention the format of the req.
Yes, that's a good point I'll update that

>  If needed, later on we can see whatever customization we need. WDYT?
As usually, it will be harder to come back to this and make changes so I would prefer not having to itarate over it if really needed tbh. I also think we could also give both options to avoid having to deal with resources and this option to just send the json data.

But since it's okay with me to start with the json param only

Reply:
I know that `ceph smb cluster` commands are different but aren't we aligned with `ceph smb apply ...`?

Reply:
Yes, we are. But `ceph smb apply` itself is aligned with a file uploader approach (not sure if that's the best way to put it) as you expect the user to do `ceph smb apply -i data.json` and have all the resources they need to apply in the `data.json` file. So you are using a declarative approach to apply your resources. 
Where as creating different endpoints for list/create/delete... is more of a imperative approach where we are explicitly stating that we want to create a cluster, or delete a cluster, this endpoints follow the imperative approach of the `ceph smb cluster create` cmds, that's why mentioned it before.

Reply:
I think we have to chose one thing and stick with it because you were doing an imperative design for the API but in the end consumes the declarative `apply_resources` method. I think the extra logic that you put there is to mostly to adapt what we are getting according to that function call isn't it? If we want to make it like `ceph smb cluster create` then why not consume the corresponding function where that might have all these logic in the first place? Otherwise mixing them both wouldn't make much sense to me tbh. Since we prefer the declarative approach for the smb, I guess let's just follow that and get the whole thing in a json object and simplify this.